src/library/scala/collection/mutable/RedBlackTree.scala

```diff
@@ -0,0 +1,521 @@
+
+    def hasNext: Boolean = nextNode ne null
+
+    def next(): R = nextNode match {
+      case null => throw new NoSuchElementException("next on empty iterator")
+      case node =>
+        nextNode = successor(node)
+        setNullIfAfterEnd()
+        nextResult(node)
+    }
+
+    private[this] var nextNode: Node[A, B] = start match {
+      case None => minNode(tree.root)
+      case Some(from) => minNodeAfter(tree.root, from)
+    }
+
+    private[this] def setNullIfAfterEnd(): Unit =
+      if (end.isDefined && (nextNode ne null) && ord.compare(nextNode.key, end.get) >= 0)
+        nextNode = null
+
+    setNullIfAfterEnd()
+  }
+
+  private[this] final class EntriesIterator[A: Ordering, B](tree: Tree[A, B], start: Option[A], end: Option[A])
+    extends TreeIterator[A, B, (A, B)](tree, start, end) {
+
+    def nextResult(node: Node[A, B]) = (node.key, node.value)
+  }
+
+  private[this] final class KeysIterator[A: Ordering, B](tree: Tree[A, B], start: Option[A], end: Option[A])
+    extends TreeIterator[A, B, A](tree, start, end) {
+
+    def nextResult(node: Node[A, B]) = node.key
+  }
+
+  private[this] final class ValuesIterator[A: Ordering, B](tree: Tree[A, B], start: Option[A], end: Option[A])
+    extends TreeIterator[A, B, B](tree, start, end) {
+
+    def nextResult(node: Node[A, B]) = node.value
+  }
+
+  // ---- debugging ----
+
+  /**
+   * Checks if the tree is in a valid state. That happens if:
+   * - It is a valid binary search tree;
+   * - All red-black properties are satisfied;
+   * - All non-null nodes have their `parent` reference correct;
+   * - The size variable in `tree` corresponds to the actual size of the tree.
+   */
+  def isValid[A: Ordering, B](tree: Tree[A, B]): Boolean =
+    isValidBST(tree.root) && isBlack(tree.root) && blackHeight(tree.root) >= 0 && size(tree.root) == tree.size
+
+  /**
+   * Returns true if this node follows the properties of a binary search tree.
+   */
+  private[this] def isValidBST[A, B](node: Node[A, B])(implicit ord: Ordering[A]): Boolean = {
+    if (node eq null) true
+    else {
+      if ((node.left ne null) && (ord.compare(node.key, node.left.key) <= 0) ||
+        (node.right ne null) && (ord.compare(node.key, node.right.key) >= 0)) false
+      else isValidBST(node.left) && isValidBST(node.right)
+    }
+  }
+
+  /**
+   * If all paths to descendent leaves of this node have the same number of black nodes, returns that number. If not,
+   * returns -1. `null` nodes count as black nodes.
+   */
+  private[this] def blackHeight[A, B](node: Node[A, B]): Int = {
+    if (node eq null) 1
+    else if (node.red && (isRed(node.left) || isRed(node.right))) -1
+    else if ((node.left ne null) && (node.left.parent ne node)) -1
+    else if ((node.right ne null) && (node.right.parent ne node)) -1
+    else {
+      val lh = blackHeight(node.left)
+      if (lh == -1) -1
+      else {
+        val bh = blackHeight(node.right)
```

Comment:
The logic here is really simple, but I think `bh` should be `rh` so it's symmetric?  Also, since shortcutting isn't really important (this isn't for performance!), wouldn't it be clearest to write it as

```
val lh = blackHeight(node.left)
val rh = blackHeight(node.right)
if (lh == -1 || lh != rh) -1
else if (isRed(node)) lh
else lh + 1
```

or something like that?